core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -3766,6 +3771,115 @@ private static void fix(List<RexNode> operands, int before, int after) {
     }
   }
 
+  /**
+   * Gets a shuttle to expand RexLocalRef.
+   */
+  public static RexShuttle getExpandShuttle(RexProgram rexProgram) {
+    return new RexShuttle() {
+      @Override public RexNode visitLocalRef(RexLocalRef localRef) {
+        return rexProgram.expandLocalRef(localRef);
+      }
+    };
+  }
+
+  /**
+   * Determine whether there is mapping between project input and output fields.
+   * Bail out if sort relies on non-trivial expressions.
+   */
+  private static boolean isCollationOnTrivialExpr(
+      List<RexNode> projects,RelDataTypeFactory typeFactory,
+      Mappings.TargetMapping map, RelFieldCollation fc, boolean passDown) {
+    final int index = fc.getFieldIndex();
+    int target = map.getTargetOpt(index);
```

Comment:
I am not sure whether RelOptUtil is the best place to hold these methods. Any suggestion is appreciated.

Reply:
I think the part of construct pairs of traits for `passThroughTraits` and `dervieTratis` are better kept in Enumerable nodes, at least for better readbility.